spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java

```diff
@@ -226,7 +244,21 @@ public void setupListenerContainer(MessageListenerContainer listenerContainer, M
 	private void setupMessageListener(MessageListenerContainer container, MessageConverter messageConverter) {
 		MessageListener<K, V> messageListener = createMessageListener(container, messageConverter);
 		Assert.state(messageListener != null, "Endpoint [" + this + "] must provide a non null message listener");
-		container.setupMessageListener(messageListener);
+		if (this.recordFilterStrategy != null) {
+			if (messageListener instanceof AcknowledgingMessageListener) {
+				@SuppressWarnings("unchecked")
+				AcknowledgingMessageListener<K, V> aml = (AcknowledgingMessageListener<K, V>) messageListener;
+				aml = new FilteringAcknowledgingMessageListenerAdapter<>(this.recordFilterStrategy, aml, this.ackDiscarded);
+				container.setupMessageListener(aml);
+			}
+			else {
+				messageListener = new FilteringMessageListenerAdapter<>(this.recordFilterStrategy, messageListener);
+				container.setupMessageListener(messageListener);
+			}
+		}
+		else {
+			container.setupMessageListener(messageListener);
```

Comment:
We can still have this line of code without duplication in each `if...else`.
Minor, of course.
Will polish on merge.